src/parser/jobs/ast/modules/Divination.tsx

```diff
@@ -0,0 +1,297 @@
+	private castHook?: EventHook<Events['action']>
+
+	private otherAst: {
+		timeOverwritten: number,
+		countOverwritten: number,
+		active: boolean,
+		start: number
+	} = {
+		timeOverwritten: 0,
+		countOverwritten: 0,
+		active: false,
+		start: 0,
+	}
+
+	override initialise() {
+		const divinationFilter = filter<Event>().status(this.data.statuses.DIVINATION.id)
+
+		this.addEventHook(divinationFilter.source(this.parser.actor.id), this.countDivinationBuffs)
+		this.addEventHook(divinationFilter.type('statusApply')
+			.target(this.parser.actor.id), this.tryOpenWindow)
+		this.addEventHook(divinationFilter.type('statusRemove')
+			.target(this.parser.actor.id), this.tryCloseWindow)
+
+		this.addEventHook('complete', this.onComplete)
+	}
+
+	private countDivinationBuffs(event: Events['statusApply']) {
+		if (this.currentWindow != null && !this.currentWindow.playersBuffed.includes(event.target) && this.actors.get(event.target).playerControlled) {
+			this.currentWindow.playersBuffed.push(event.target)
+		}
+	}
+
+	private tryOpenWindow(event: Events['statusApply']) {
+		//open window for other AST
+		if (this.parser.actor.id !== event.source) {
+			this.otherAst.active = true
+			this.otherAst.start = event.timestamp
+			//if window was open. close it prematurely :(
+			if (this.currentWindow != null) {
+				const eventClose: Events['statusRemove'] = {
+					timestamp: event.timestamp,
+					status: this.data.statuses.DIVINATION.id,
+					type: 'statusRemove',
+					source: event.source,
+					target: this.parser.actor.id,
+				}
+				this.tryCloseWindow(eventClose)
+			}
+		}
+
+		if (this.currentWindow === undefined && this.parser.actor.id === event.source) {
+			//open window for ast
+			this.currentWindow = {
+				start: event.timestamp,
+				rotation: [],
+				gcdCount: 0,
+				source: event.source,
+
+				playersBuffed: [],
+				overlap: false,
+			}
+			this.castHook = this.addEventHook(
+				filter<Event>()
+					.source(this.parser.actor.id)
+					.type('action'),
+				this.onCast,
+			)
+
+			//to set other ast information if active and overwritten
+			//note assumption that overwritten happens if duration hasn't passed
+			if (this.data.statuses.DIVINATION.duration > this.currentWindow.start - this.otherAst.start) {
+				this.otherAst.countOverwritten++
+				this.otherAst.timeOverwritten += (this.otherAst.start + this.data.statuses.DIVINATION.duration) - event.timestamp
+				this.currentWindow.overlap = true
+			}
+			this.otherAst.active = false
+			this.otherAst.start = 0
+		}
+	}
+
+	private tryCloseWindow(event: Events['statusRemove']) {
+		//note: closing other AST window handled in open window since closing the window doesn't catch what we need because status remove happens before status apply
+
+		//to stop using function if window isn't even open. used when splicing
+		if (this.currentWindow == null) {
+			return
+		}
+
+		//close window and push to history
+		if (event.source === this.parser.actor.id) {
```

Comment:
I think it would be better to limit this function to checking if we need to close the window here based on the source, and then passing the timestamp to a second function that actually does close the window.  Then the "close when your window gets overwritten" handler up in tryOpenWindow could just pass the timestamp to the actual close function as well and you don't have to synth a full event for it, while being a little more clear about why you're only trying to close certain windows.  (And similarly onComplete wouldn't have to synth an event either.  You could also just use this.parser.currentTimestamp in the actual close function and never pass around the event.timestamp).

Reply:
added function to do as described and updated the relating splices as a result